File: q1/q1d.py
# Filename: q1d.py
# Date Created: 13-Mar-2019 10:04:58 pm
# Description:
import pandas as pd
from math import erf, sqrt, exp, log, pow, pi
import matplotlib.pyplot as plt
import numpy as np
from functions import *
pd.set_option('display.max_columns', 20)

# Load in the outputs from q1b.py
# Processed actual option prices
df_calls = pd.read_pickle("./outputs/df_calls.pkl")
df_puts = pd.read_pickle("./outputs/df_puts.pkl")
FTSE100 = pd.read_pickle("./outputs/FTSE100.pkl")

# Estimated volatility for call
est_vol_df = pd.read_pickle("./outputs/est_vol_df.pkl")

# Computed BS option prices
bs_call_df = pd.read_pickle("./outputs/bs_call_df.pkl")
bs_put_df = pd.read_pickle("./outputs/bs_put_df.pkl")

# Find the

# Strike price
strike_price = '7300'
K = int(strike_price)

# date - 06/07/2018
t = 135
ttm = (277 - t ) / 365 # expressed in years
ttm
# Estimated volatility
annual_vol = est_vol_df[strike_price][t]

# Risk free interest rate
r = FTSE100.iloc[t-1,2] / 100

# Stock price on that date
S = FTSE100.iloc[t-1,1]

# Find the Call option price
d_1 = d1(S,K,r,annual_vol,ttm)
d_2 = d2(d_1,annual_vol,ttm)

# Compute the call option price
c_price = ComputeCallOptionPrice(S,K,d_1,d_2,ttm,r)

c_price
S
annual_vol
r
FTSE100.iloc[t-1,1]

df_calls[strike_price][t]
FTSE100.iloc[t-1,0]

File: q1/q1d.m
% Load in the specific dataset
% TTM (expressed in years)
T = (277 - 135) / 365;

% Strike Price
K = 7300;

% Risk free interest rate
r = 0.01267;

% Stock price on that date
S0 = 7617.7;

% Estimated volatility
sigma = 0.11412013507029374;

% BS output from python
BS_price = 433.4412656706718;

% Compute the Binomial Lattice output for different N
disp('Price     deltaT');
lattice_prices = [];
deltaT_lst = [];

% N - No of time steps
for N=4:5
    [price, lattice] = LatticeEurCall(S0,K,r,T,sigma,N);
    deltaT = T/N;
    lattice_prices = [lattice_prices, price];
    deltaT_lst = [deltaT_lst, deltaT];
end

% Find the absolute difference between the 2 methods
abs_diff = abs(lattice_prices - BS_price);

% Plot the absolute diferrence between the 2 method against dt.
plot(deltaT_lst, abs_diff);
ylabel('Absolute Difference');
xlabel('Step Time, \deltat');
grid();


% The actual price of the market
actual_price = 402;


% for tau=1:N
%     for i= (tau+1):2:(2*N+1-tau)
%         disp([i, tau]);
%     end
% end
% 
% 
% for i=1:2:2*N+1
%     disp(i);
% end